File: shared-sale/src/main/java/com/elitec/shared/sale/feature/sale/domain/entity/Sale.kt
package com.elitec.shared.sale.feature.sale.domain.entity

import kotlinx.datetime.LocalDate
import kotlinx.serialization.Serializable

@Serializable
data class Sale(
    val id: String,
    val date: LocalDate,
    val amount: Double,
    val currency: Currency,
    val verified: BuyState,
    val products: List<SaleItem>,
    val userId: String,
    val customerName: String? = null,
    val deliveryType: DeliveryType? = null,
    val deliveryAddress: DeliveryAddress? = null,
    /**
     * SALE_POLICY: tipo comercial al cerrar en operador.
     * null mientras UNVERIFIED (provisional); obligatorio al pasar a VERIFIED.
     */
    val saleType: SaleType? = null
)

enum class BuyState {
    UNVERIFIED, VERIFIED, DELETED
}

/**
 * SALE_POLICY Core 1 — tipos de venta definidos en tienda.
 * Afectan importe; no eximen de baja de stock al confirmar.
 */
enum class SaleType {
    /** Precio de lista del catálogo. */
    NORMAL,
    /** Descuento alineado con dueño o promoción. */
    DISCOUNT,
    /** Regalia / obsequio — amount 0, stock sí baja. */
    GIFT
}

/**
 * Preferencia de entrega que el cliente elige cuando su pedido
 * está en estado VERIFIED (listo para entrega).
 */
enum class DeliveryType {
    PICKUP,   // El cliente va a recogerlo al taller
    DELIVERY  // El cliente solicita domicilio (el taller coordina)
}

enum class Currency { CUP, USD, MLC }

@Serializable
data class DeliveryAddress(
    val province: String,
    val municipality: String,
    val mainStreet: String,
    val betweenStreets: String? = null,
    val phone: String,
    val houseNumber: String,
    val referenceName: String? = null
)

File: web/src/core/feature/sale/domain/entity/Sale.ts
import {type BuyState, Currency, DeliveryType, type SaleType} from "./enums";

export interface Sale {
    id: string
    date: string // ISO string (mejor que LocalDate en frontend)
    amount: number
    verified: BuyState
    products: SaleItem[]
    currency: Currency
    userId: string
    deliveryType?: DeliveryType | null
    deliveryAddress?: DeliveryAddress | null
    /** SALE_POLICY: null mientras UNVERIFIED; obligatorio en VERIFIED */
    saleType?: SaleType | null
    /**
     * Soft-hold aplicado: reserved++ por cada línea al crear UNVERIFIED.
     * Evita doble hold en reintentos (idempotencia).
     */
    stockHoldApplied?: boolean
}

export interface SaleItem {
    productId: string
    productName?: string | null
    quantity: number
    /** Precio unitario efectivo (lista, descuento o 0 si GIFT) */
    unitPrice?: number | null
    /** Precio de lista al momento (auditoría) */
    listUnitPrice?: number | null
    /** @deprecated prefer unitPrice; mantenido por compat con payloads antiguos */
    price?: number
}

export interface DeliveryAddress {
    province: string
    municipality: string
    mainStreet: string
    betweenStreets?: string | null
    phone: string
    houseNumber: string
    referenceName?: string | null
}
